File: examples/qwen2/main.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include <ATen/ATen.h>

#include "dli/cuda_runtime.h"
#include "dli/engine.h"
#include "dli/graph.h"
#include "dli/weights.h"

namespace {

struct Args {
  std::string graph = "build/examples/qwen2/qwen2.dli.json";
  std::string weights = "build/examples/qwen2/qwen2.dli.weights.json";
  std::string plugin = "build/operators/libdli_triton_aot_ops.so";
};

Args parseArgs(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];
    if (arg == "--graph" && i + 1 < argc) args.graph = argv[++i];
    else if (arg == "--weights" && i + 1 < argc) args.weights = argv[++i];
    else if (arg == "--plugin" && i + 1 < argc) args.plugin = argv[++i];
    else throw std::invalid_argument("usage: dli_qwen2 [--graph path] [--weights path] [--plugin path]");
  }
  return args;
}

dli::Tensor torchCudaInt64(std::vector<std::int64_t> shape, const std::vector<std::int64_t>& values) {
  auto tensor = at::tensor(values, at::TensorOptions().dtype(at::kLong)).view(shape);
  return dli::Tensor(tensor.to(at::Device(at::kCUDA, 0)));
}

std::vector<float> downloadFloat(const dli::Tensor& tensor) {
  std::vector<float> values(tensor.numel());
  dli::cudaMemcpyBytes(values.data(), tensor.deviceData(), tensor.nbytes(), dli::CudaMemcpyKind::DeviceToHost);
  return values;
}

}  // namespace

int main(int argc, char** argv) {
  try {
    const auto args = parseArgs(argc, argv);
    dli::Engine engine;
    engine.registry().loadLibrary(args.plugin);
    auto tensors = dli::loadWeights(args.weights, dli::DeviceType::Cuda);
    tensors["input_ids"] = torchCudaInt64({1}, {2});
    auto outputs = engine.run(dli::Graph::fromJsonFile(args.graph), std::move(tensors));
    const auto values = downloadFloat(outputs.at("logits"));
    std::cout << "logits:";
    for (std::size_t i = 0; i < std::min<std::size_t>(values.size(), 16); ++i) std::cout << " " << values[i];
    std::cout << "\nkv_cache_entries: " << engine.kvCache().size() << "\n";
    return 0;
  } catch (const std::exception& error) {
    std::cerr << "qwen2 example failed: " << error.what() << "\n";
    return 1;
  }
}

File: examples/qwen2/main.py
from __future__ import annotations

import argparse
import sys
from pathlib import Path


def _repo_root(start: Path) -> Path:
    for path in [start, *start.parents]:
        if (path / "CMakeLists.txt").exists() and (path / "python").exists():
            return path
    raise RuntimeError("failed to locate repository root")


ROOT = _repo_root(Path(__file__).resolve())
sys.path.insert(0, str(ROOT / "build" / "python"))
sys.path.insert(1, str(ROOT / "python"))

import torch

import dli
from dli_export import export_module
from qwen2 import create_model


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--build-dir", type=Path, default=ROOT / "build")
    parser.add_argument("--output-dir", type=Path)
    parser.add_argument("--plugin", type=Path)
    parser.add_argument("--model-id")
    parser.add_argument("--local-files-only", action="store_true")
    parser.add_argument("--export-only", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    output_dir = args.output_dir or args.build_dir / "examples" / "qwen2"
    plugin = args.plugin or args.build_dir / "operators" / "libdli_triton_aot_ops.so"

    graph_path, weights_path = export_module(
        create_model(model_id=args.model_id, local_files_only=args.local_files_only),
        (),
        output_dir,
        model_type="qwen2",
        stem="qwen2",
    )
    print(f"graph: {graph_path}")
    print(f"weights: {weights_path}")
    if args.export_only:
        return 0
    if not torch.cuda.is_available():
        raise RuntimeError("qwen2 example requires CUDA because the AOT operators run on GPU")

    inputs = dli.load_weights(str(weights_path), "cuda")
    inputs["input_ids"] = torch.tensor([2], dtype=torch.int64, device="cuda")

    engine = dli.Engine()
    engine.load_library(str(plugin))
    outputs = engine.run(dli.Graph.from_json_file(str(graph_path)), inputs)
    torch.cuda.synchronize()
    logits = outputs["logits"].detach().cpu().flatten()[:16].tolist()
    print("logits:", " ".join(str(value) for value in logits))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
